facebook/delphiFacebook/R/contingency_run.R

```diff
@@ -0,0 +1,121 @@
+#'   to report. `name` is the aggregate's base column name; `var_weight` is the 
+#'   column to use for its weights; `metric` is the column of `df` containing the
+#'   response value. `group_by` is a list of variables used to perform the 
+#'   aggregations over. `compute_fn` is the function that computes
+#'   the aggregate response given many rows of data. `post_fn` is applied to the
+#'   aggregate data after megacounty aggregation, and can perform any final
+#'   calculations necessary.
+#'
+#' @return none
+#' 
+#' @importFrom lubridate ymd days
+#' 
+#' @export
+run_contingency_tables <- function(params, aggregations)
+{
+  if (!is.null(params$end_date) & !is.null(params$n_periods)) {
+    ## Produce historical CSVs
+    
+    if (params$aggregate_range == "week") {
+      period_step <- days(7)
+    } else if (params$aggregate_range == "week") {
+      period_step <- months(1)
+    } else if (is.null(params$aggregate_range)) {
+      stop("setting aggregate_range must be provided in params")
+    }
+    
+    # Make list of dates to aggregate over.
+    end_dates <- sort( ymd(params$end_date) - period_step * seq(0, params$n_periods) )
+    
+    for (end_date in end_dates) {
+      period_params <- params
+      period_params$end_date <- end_date
+      run_contingency_tables_one_period(period_params, aggregations)
+    }
+  } else {
+    ## Produce CSVs for a single time period
+    run_contingency_tables_one_period(params, aggregations)
+  }
+  
+
+}
+
+#' Run the contingency table production pipeline for a single time period
+#'
+#' See the README.md file in the source directory for more information about how to run
+#' this function.
+#'
+#' @param params    Params object produced by read_params
+#' @param aggregations Data frame with columns `name`, `var_weight`, `metric`,
+#'   `group_by`, `compute_fn`, `post_fn`. Each row represents one aggregate
+#'   to report. `name` is the aggregate's base column name; `var_weight` is the 
+#'   column to use for its weights; `metric` is the column of `df` containing the
+#'   response value. `group_by` is a list of variables used to perform the 
+#'   aggregations over. `compute_fn` is the function that computes
+#'   the aggregate response given many rows of data. `post_fn` is applied to the
+#'   aggregate data after megacounty aggregation, and can perform any final
+#'   calculations necessary.
+#'
+#' @return none
+#' 
+#' @importFrom parallel detectCores
+#' 
+#' @export
+run_contingency_tables_one_period <- function(params, aggregations)
+{
+  params <- update_params(params)
+  aggregations <- verify_aggs(aggregations)
+  
+  cw_list <- produce_crosswalk_list(params$static_dir)
+  archive <- load_archive(params)
+  msg_df("archive data loaded", archive$input_data)
+  
+  input_data <- load_responses_all(params)
+  input_data <- filter_responses(input_data, params)
+  input_data <- bodge_v4_translation(input_data)
+  msg_df("response input data", input_data)
```

Comment:
I wonder if we should move this ingestion sequence into a function so that if we add additional steps, that can be done in one place

(and maybe we should factor out `bodge_v4_translation` to happen in one of the other functions? it's only in `create_complete_responses` because none of the aggregates need it, but if the contingency tables will need V4, may as well put it in `load_responses_all` or something)

Reply:
Actually I'm going to move `bodge_v4_translation` into `load_responses_one`, above where all the variables are coded, because I need to code V4 to put it into the API. So you'll just need to handle this change when merging/rebasing on `fb-package` once I'm done

Reply:
I pushed the change moving `bodge_v4_translation`, so you should be able to get rid of this now